File: 2015/21/solution.ts
import * as fs from "fs";
import * as path from "path";

const puzzleInput = fs.readFileSync(path.join(__dirname, "input.txt"), "utf-8");

const shopInput = `Weapons:    Cost  Damage  Armor
Dagger        8     4       0
Shortsword   10     5       0
Warhammer    25     6       0
Longsword    40     7       0
Greataxe     74     8       0

Armor:      Cost  Damage  Armor
Leather      13     0       1
Chainmail    31     0       2
Splintmail   53     0       3
Bandedmail   75     0       4
Platemail   102     0       5

Rings:      Cost  Damage  Armor
Damage +1    25     1       0
Damage +2    50     2       0
Damage +3   100     3       0
Defense +1   20     0       1
Defense +2   40     0       2
Defense +3   80     0       3`
  .split("\n\n")
  .map((a) => a.split("\n"));

class Character {
  hp: number;
  damage: number;
  armor: number;
  cost: number;
  constructor(equipment: Item[]) {
    this.hp = 0;
    this.damage = 0;
    this.armor = 0;
    this.cost = 0;
    for (const item of equipment) {
      this.equip(item);
    }
  }

  equip(item: Item) {
    this.hp += item.hp;
    this.damage += item.damage;
    this.armor += item.armor;
    this.cost += item.cost;
  }

  attack(enemy: Character) {
    enemy.hp -= Math.max(this.damage - enemy.armor, 1);
  }
}

interface Item {
  name: string;
  hp: number;
  cost: number;
  damage: number;
  armor: number;
}

const itemRegex = /^(.*?)\s{2,}(\d+)\s+(\d+)\s+(\d+)$/;

function parseShopInput(shopInput: string[]): Item[] {
  const items: Item[] = [];
  for (const line of shopInput) {
    const parts = line.match(itemRegex);
    if (!parts) continue;
    items.push({
      name: parts[1],
      hp: 0,
      cost: Number(parts[2]),
      damage: Number(parts[3]),
      armor: Number(parts[4]),
    });
  }
  return items;
}

const weapons = parseShopInput(shopInput[0].slice(1));
const armors = parseShopInput(shopInput[1].slice(1));
armors.push({ name: "No Armor", hp: 0, cost: 0, damage: 0, armor: 0 });
const rings = parseShopInput(shopInput[2].slice(1));
rings.push({ name: "No Ring", hp: 0, cost: 0, damage: 0, armor: 0 });

const ringCombos: [Item, Item][] = [];
for (const ring1 of rings) {
  for (const ring2 of rings) {
    ringCombos.push([ring1, ring2]);
  }
}

const playerBase: Item = {
  name: "Player",
  hp: 100,
  cost: 0,
  damage: 0,
  armor: 0,
};

const bossStats = puzzleInput.match(/\d+/g)!.map(Number);

const bossBase: Item = {
  name: "Boss",
  hp: bossStats[0],
  cost: 0,
  damage: bossStats[1],
  armor: bossStats[2],
};

function simulate(player: Character, boss: Character): boolean {
  while (true) {
    player.attack(boss);
    if (boss.hp <= 0) return true;
    boss.attack(player);
    if (player.hp <= 0) return false;
  }
}

const scenarios: [Character, Character][] = [];

for (const weapon of weapons) {
  for (const armor of armors) {
    for (const ringCombo of ringCombos) {
      const player = new Character([playerBase, weapon, armor, ...ringCombo]);
      const boss = new Character([bossBase]);
      scenarios.push([player, boss]);
    }
  }
}

const playerWinCosts: number[] = [];
const playerLossCosts: number[] = [];

for (const [player, boss] of scenarios) {
  const playerWon = simulate(player, boss);
  if (playerWon) playerWinCosts.push(player.cost);
  else playerLossCosts.push(player.cost);
}

console.log(Math.min(...playerWinCosts), Math.max(...playerLossCosts));

File: 2015/21/solution.go
package main

import (
	"fmt"
	"math"
	"os"
	"path/filepath"
	"regexp"
	"runtime"
	"strconv"
	"strings"
)

func main() {
	_, filename, _, _ := runtime.Caller(0)
	dirname := filepath.Dir(filename)
	inputFilePath := filepath.Join(dirname, "input.txt")
	data, err := os.ReadFile(inputFilePath)
	if err != nil {
		panic(err)
	}
	puzzleInput := strings.Split(string(data), "\n")
	boss := parseInput(puzzleInput)
	shops := parseItemShop(ITEMSHOP)
	fmt.Println(solve(boss, shops))
}

const ITEMSHOP = `Weapons:    Cost  Damage  Armor
Dagger        8     4       0
Shortsword   10     5       0
Warhammer    25     6       0
Longsword    40     7       0
Greataxe     74     8       0

Armor:      Cost  Damage  Armor
Leather      13     0       1
Chainmail    31     0       2
Splintmail   53     0       3
Bandedmail   75     0       4
Platemail   102     0       5

Rings:      Cost  Damage  Armor
Damage +1    25     1       0
Damage +2    50     2       0
Damage +3   100     3       0
Defense +1   20     0       1
Defense +2   40     0       2
Defense +3   80     0       3`

type Item struct {
	name   string
	hp     int
	cost   int
	damage int
	armor  int
}

var playerBase = &Item{"Player", 100, 0, 0, 0}

type Character struct {
	Item
	equipment []*Item
}

func (ch *Character) Equip(item *Item) {
	ch.hp += item.hp
	ch.cost += item.cost
	ch.damage += item.damage
	ch.armor += item.armor
	ch.equipment = append(ch.equipment, item)
}

func (ch *Character) Attack(target *Character) {
	damage := ch.damage - target.armor
	if damage < 1 {
		damage = 1
	}
	target.hp -= damage
}

func NewCharacter(base *Item, equipment []*Item) *Character {
	ch := &Character{*base, make([]*Item, 0, len(equipment))}
	for _, item := range equipment {
		ch.Equip(item)
	}
	return ch
}

func parseItemShop(itemShop string) [][]*Item {
	shopStrings := strings.Split(itemShop, "\n\n")
	shops := make([][]*Item, len(shopStrings))
	itemRegex := regexp.MustCompile(`^(.*?)\s{2,}(\d+)\s+(\d+)\s+(\d+)$`)
	for i, part := range shopStrings {
		lines := strings.Split(part, "\n")
		items := make([]*Item, 0, len(lines)+1)
		for _, line := range lines[1:] {
			parts := itemRegex.FindStringSubmatch(line)
			if parts == nil {
				panic(fmt.Sprintf("Could not match itemRegex with: %s", line))
			}
			name := parts[1]
			hp := 0
			cost, err := strconv.Atoi(parts[2])
			if err != nil {
				panic(err)
			}
			damage, err := strconv.Atoi(parts[3])
			if err != nil {
				panic(err)
			}
			armor, err := strconv.Atoi(parts[4])
			if err != nil {
				panic(err)
			}
			items = append(items, &Item{name, hp, cost, damage, armor})
		}
		shops[i] = items
	}
	shops[1] = append(shops[1], &Item{"No Armor", 0, 0, 0, 0})
	shops[2] = append(shops[2], &Item{"No Ring", 0, 0, 0, 0})
	return shops
}

func parseInput(data []string) *Item {
	numRegex := regexp.MustCompile(`\d+`)
	hpString := numRegex.FindString(data[0])
	hp, err := strconv.Atoi(hpString)
	if err != nil {
		panic(err)
	}
	damageString := numRegex.FindString(data[1])
	damage, err := strconv.Atoi(damageString)
	if err != nil {
		panic(err)
	}
	armorString := numRegex.FindString(data[2])
	armor, err := strconv.Atoi(armorString)
	if err != nil {
		panic(err)
	}
	return &Item{"Boss", hp, 0, damage, armor}
}

func simulate(player, boss *Character) bool {
	for {
		player.Attack(boss)
		if boss.hp <= 0 {
			return true
		}
		boss.Attack(player)
		if player.hp <= 0 {
			return false
		}
	}
}

func solve(bossBase *Item, shops [][]*Item) (int, int) {
	var scenarios [][2]*Character
	weaponShop := shops[0]
	armorShop := shops[1]
	ringShop := shops[2]
	for _, weapon := range weaponShop {
		for _, armor := range armorShop {
			for _, ring1 := range ringShop {
				for _, ring2 := range ringShop {
					player := NewCharacter(playerBase, []*Item{weapon, armor, ring1, ring2})
					boss := NewCharacter(bossBase, []*Item{})
					scenarios = append(scenarios, [2]*Character{player, boss})
				}
			}
		}
	}

	part1 := math.MaxInt
	part2 := 0
	for _, scenario := range scenarios {
		player, boss := scenario[0], scenario[1]
		playerWon := simulate(player, boss)
		if playerWon {
			if player.cost < part1 {
				part1 = player.cost
			}
		} else {
			if player.cost > part2 {
				part2 = player.cost
			}
		}
	}

	return part1, part2
}
